docs/sql/02-type-system.md

```diff
@@ -0,0 +1,175 @@
+
+*Table 2: SQL-to-Java mapping*
+
+| SQL Type          | Java Type                  |
+|-------------------|----------------------------|
+| `NULL`            | `java.lang.Void`           |
+| `VARCHAR`         | `java.lang.String`         |
+| `BOOLEAN`         | `java.lang.Boolean`        |
+| `TINYINT`         | `java.lang.Byte`           |
+| `SMALLINT`        | `java.lang.Short`          |
+| `INTEGER`         | `java.lang.Integer`        |
+| `BIGINT`          | `java.lang.Long`           |
+| `DECIMAL`         | `java.math.BigDecimal`     |
+| `REAL`            | `java.lang.Float`          |
+| `DOUBLE`          | `java.lang.Double`         |
+| `DATE`            | `java.time.LocalDate`      |
+| `TIME`            | `java.time.LocalTime`      |
+| `TIMESTAMP`       | `java.time.LocalDateTime`  |
+| `TIMESTAMP W/ TZ` | `java.time.OffsetDateTime` |
+| `OBJECT`          | `java.lang.Object`         |
+
+`TIMESTAMP WITH TIME ZONE` is mapped to the `java.time.OffsetDateTime` class because ANSI SQL requires only zone
+displacement, so full zone information from the `java.time.ZonedDateTime` class is not needed.
+
+### 2.2 Java to SQL Mapping
+
+Table 3 establishes a many-to-one mapping between Java and SQL types.
+
+*Table 3: Java-to-SQL mapping*
+
+| Java Type                  | SQL Type          |
+|----------------------------|-------------------|
+| `java.lang.Void`           | `NULL`            |
+| `java.lang.String`         | `VARCHAR`         |
+| `java.lang.Character`      | `VARCHAR`         |
+| `java.lang.Boolean`        | `BOOLEAN`         |
+| `java.lang.Byte`           | `TINYINT`         |
+| `java.lang.Short`          | `SMALLINT`        |
+| `java.lang.Integer`        | `INTEGER`         |
+| `java.lang.Long`           | `BIGINT`          |
+| `java.math.BigInteger`     | `DECIMAL`         |
+| `java.math.BigDecimal`     | `DECIMAL`         |
+| `java.lang.Float`          | `REAL`            |
+| `java.lang.Double`         | `DOUBLE`          |
+| `java.time.LocalDate`      | `DATE`            |
+| `java.time.LocalTime`      | `TIME`            |
+| `java.time.LocalDateTime`  | `TIMESTAMP`       |
+| `java.util.Calendar`       | `TIMESTAMP W/ TZ` |
+| `java.util.Date`           | `TIMESTAMP W/ TZ` |
+| `java.time.Instant`        | `TIMESTAMP W/ TZ` |
+| `java.time.OffsetDateTime` | `TIMESTAMP W/ TZ` |
+| `java.time.ZonedDateTime`  | `TIMESTAMP W/ TZ` |
+| Any other type             | `OBJECT`          |
+
+The following SQL types are mapped to several Java types:
+
+- `VARCHAR` is mapped to `java.lang.String` and `java.lang.Character`
+- `DECIMAL` is mapped to `java.math.BigInteger` and `java.math.BigDecimal`
+- `TIMESTAMP W/ TZ` is mapped to multiple date/time classes which represent a time instant.
+
+## 3. Type Conversions
+
+Different types might be converted to each other. The table provides the list of type conversions.
+
+*Table 4: Type conversions (I - implicit, E - explicit)*
```

Comment:
* legend does not match the table which contains `Y`.
* I think there are different rules for operators (and row conversions) and different for functions, and maybe also different for dynamic parameters
* lack of JSON (which is tricky and changed in #22746)